src/tests/ftest/deployment/server_rank_failure.py

```diff
@@ -0,0 +1,378 @@
+        else:
+            self.log.info("daos_engine in %s killed", engine_kill_host)
+
+    def verify_ior_worked(self, ior_results, job_num, errors):
+        """Verify that the IOR worked.
+
+        Args:
+            ior_results (dict): Dictionary that contains the IOR run results.
+            job_num (int): Job number used for the IOR run.
+            errors (list): Error list used in the test.
+        """
+        self.log.info(ior_results)
+        if not ior_results[job_num][0]:
+            ior_error = ior_results[job_num][1]
+            errors.append("Error found in second IOR run! {}".format(ior_error))
+
+    def verify_rank_failure(self, ior_namespace, container_namespace):
+        """Verify engine failure can be recovered by restarting daos_server.
+
+        1. Create a pool and a container. Create a container with or without redundancy
+        factor based on container_namespace.
+        2. Run IOR with given object class.
+        3. While IOR is running, kill daos_engine processes from one of the nodes to
+        simulate a node failure. (Two engines per node.)
+        4. Verify that IOR failed.
+        5. Restart daos_server service on the two nodes.
+        6. Verify the system status by calling dmg system query.
+        7. Verify that the container Health is HEALTHY.
+        8. Run IOR to the same container and verify that it works.
+
+        Args:
+            ior_namespace (str): Yaml namespace that defines the object class used for
+                IOR.
+            container_namespace (str): Yaml namespace that defines the container
+                redundancy factor.
+        """
+        # 1. Create a pool and a container.
+        self.add_pool(namespace="/run/pool_size_ratio_80/*")
+        self.add_container(pool=self.pool, namespace=container_namespace)
+
+        # 2. Run IOR.
+        ior_results = {}
+        job_num = 1
+        # If we don't use timeout, when the engines are killed, Mpirun gets stuck and
+        # waits forever. If we use timeout that's too long, the daos container command
+        # after the server restart will get stuck, so use 10 sec timeout, which is the
+        # same as deadlineForStonewalling in IOR.
+        mpirun_timeout = self.params.get('sw_deadline', ior_namespace)
+        self.log.info(
+            "Running Mpirun-IOR with Mpirun timeout of %s sec", mpirun_timeout)
+        ior_thread = threading.Thread(
+            target=self.run_ior_report_error,
+            args=[ior_results, job_num, "test_file_1", self.pool, self.container,
+                  ior_namespace, mpirun_timeout])
+
+        ior_thread.start()
+
+        # Wait for a few seconds for IOR to start.
+        self.log.info("Waiting 5 sec for IOR to start writing data...")
+        time.sleep(5)
+
+        # 3. While IOR is running, kill daos_engine from two of the ranks.
+        # Arbitrarily select node index, say index 1.
+        self.kill_engine(engine_kill_host=self.hostlist_servers[1])
```

Comment:
It was discussed in one of the previous failure test PRs, that we should identify first to which rank the IO is going and then stop/kill that rank. That can be achieved using telemetry data. I think we should incorporate it. This might be working right now when there are only 3 server nodes but when dealing with hundreds of servers with 2 ranks each, this test might not work.

Reply:
I think that's a good point and I'm not against implementing it, but after doing some research, I'm not sure how to check it. I believe the test would fail if we kill the engine where there's no IO, so I think this is okay for now, unless you have some idea.

Reply:
I implemented the fix using `engine_pool_ops_update`. Now it determines which node's engines to kill from this telemetry data. Thanks.